pkg/urbit/vere/saga.c

```diff
@@ -424,17 +424,83 @@ u3_saga_open(const c3_path* const pax_u, u3_meta* const met_u)
   }
 
   try_list(log_u->epo_u.lis_u = c3_list_init(), goto free_dir_entries);
-  u3_epoc *poc_u, *pre_u;
+  u3_epoc *poc_u;
   c3_w* lif_w = &met_u->lif_w;
+  c3_d  nex_d = 0;
+  c3_l  mex_l = 0;
   for ( size_t idx_i = 0; idx_i < ent_i; idx_i++ ) {
     c3_path_push(log_u->pax_u, ent_c[idx_i]);
     try_epoc(poc_u = u3_epoc_open(log_u->pax_u, lif_w), goto free_dir_entries);
     c3_path_pop(log_u->pax_u);
+
+    if ( NULL != san_i ) {
+      //  check epoch's event range for contiguousness with earlier epochs.
+      //  first epoch may start at arbitrary event nr, so we don't check it.
+      //
+      c3_d fir_d = u3_epoc_first_commit(poc_u);
+      if ( (0 != nex_d) && (nex_d != fir_d) ) {
+        u3l_log("sane: epoch snapshot not contiguous: "
+                "expected %" PRIu64 ", got %" PRIu64 ", in %s\n",
+                nex_d, fir_d, ent_c[idx_i]);
+        *san_i = 1;
```

Comment:
I'm not a fan of the status codes because they don't seem to add much additional information, especially considering that they're undocumented. I'd be inclined to rely on the `NULL`/not-`NULL` state of the return value to determine whether the event log was sane (if not sane, the return value will be `NULL`) and allow the log messages to communicate *why* the event log was not sane.